src/meta/processors/admin/Balancer.cpp

```diff
@@ -424,30 +466,74 @@ bool Balancer::balanceParts(BalanceID balanceId,
         return false;
       }
 
-      partsFrom.erase(it);
       if (std::find(partsTo.begin(), partsTo.end(), partId) != partsTo.end()) {
         LOG(ERROR) << "Part " << partId << " already existed in partsTo";
         return false;
       }
 
+      if (dependentOnGroup) {
+        auto& parts = relatedParts_[minPartsHost.first];
+        if (!checkZoneLegal(maxPartsHost.first, minPartsHost.first) &&
+            std::find(parts.begin(), parts.end(), partId) != parts.end()) {
+          LOG(INFO) << "Zone have exist part: " << partId;
+          continue;
+        }
+      }
+
+      partsFrom.erase(it);
       partsTo.emplace_back(partId);
       tasks.emplace_back(
           balanceId, spaceId, partId, maxPartsHost.first, minPartsHost.first, kv_, client_);
       noAction = false;
     }
+
     if (noAction) {
       LOG(INFO) << "Here is no action";
       break;
     }
     sortedHosts = sortedHostsByParts(confirmedHostParts);
     maxPartsHost = sortedHosts.back();
     minPartsHost = sortedHosts.front();
+    if (innerBalance_) {
+      LOG(INFO) << "maxPartsHost.first " << maxPartsHost.first << " minPartsHost.first "
+                << minPartsHost.first;
+      while (!checkZoneLegal(maxPartsHost.first, minPartsHost.first)) {
+        sortedHosts.pop_back();
+        maxPartsHost = sortedHosts.back();
+      }
+
+      auto& source = maxPartsHost.first;
+      auto iter = std::find_if(zoneParts_.begin(), zoneParts_.end(), [&source](const auto& pair) {
+        return source == pair.first;
+      });
+
+      auto& zoneName = iter->second.first;
+      int32_t hostsSize = zoneHosts_[zoneName].size();
+      int32_t totalPartsZone = 0;
+      for (auto& host : zoneHosts_[zoneName]) {
+        auto it = confirmedHostParts.find(host);
+        if (it == confirmedHostParts.end()) {
+          LOG(ERROR) << "Host " << host << "not in confirmedHostParts";
+          continue;
+        }
+        totalPartsZone += it->second.size();
+      }
+
+      LOG(INFO) << "Update min and max loading";
+      LOG(INFO) << "Total parts in zone " << totalPartsZone << ", total hosts " << hostsSize;
+      avgLoad = static_cast<float>(totalPartsZone) / hostsSize;
+      LOG(INFO) << "The expect avg load is " << avgLoad;
+      minLoad = std::floor(avgLoad);
+      maxLoad = std::ceil(avgLoad);
```

Comment:
From line 497 to 527 , these codes just for one LOG (line 528), right?

Reply:
merge logs into one line